mobile/library/jni/jni_helper.h

```diff
@@ -141,9 +141,27 @@ class JniHelper {
   /** Gets the JNI version supported. */
   static jint getVersion();
 
-  /** Initializes the `JavaVM`. This is typically set in `JNI_OnLoad`. */
+  /** Initializes the `JavaVM`. This function is typically called inside `JNI_OnLoad`. */
   static void initialize(JavaVM* java_vm);
 
+  /**
+   * Adds the `jclass` object into a cache. This function is typically called inside `JNI_OnLoad`.
+   *
+   * Caching the `jclass` can be useful for performance.
+   * See https://developer.android.com/training/articles/perf-jni#jclass,-jmethodid,-and-jfieldid
+   *
+   * Another reason for caching the `jclass` object is to able to find a non-built-in class when the
+   * native code creates a thread and then attaching it with `AttachCurrentThread`, i.e. calling
+   * `getThreadLocalEnv()->getEnv()->FindClass`. This is because there are no stack frames from the
+   * application. When calling `FindClass` from the thread, the `JavaVM` will start in the "system"
```

Comment:
Thank you for this comment, very helpful to know!  One thing that wasn't clear is what stack frames have to do with the reset of the explanation about `FindClass`?

Reply:
From https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class

If the `JNIEnv` is attached to the current thread, there's no stack frame and `FindClass` needs the stack frame to find the class. For built-in classes, e.g. `java.util.*`, etc this is fine because we look in the system classloader, but for non-built-in classes, this will result in a class not found. Specific to this PR, we need to find `EnvoyStreamIntel` and `EnvoyFinalStreamIntel`, which are custom classes. In the next PR, I will update the rest of the code to cache the `jclass`-es since that's also the recommendation from https://developer.android.com/training/articles/perf-jni#jclass,-jmethodid,-and-jfieldid

Reply:
Another note for this, on Android each app has a separate `ClassLoader` (for sandboxing purposes). The old code does this: https://github.com/envoyproxy/envoy/blob/266f9a35ab9d698ec807999c7d5bfd226bdeb8bf/mobile/library/jni/jni_utility.cc#L24-L33

It basically tries to read any custom code from the app `ClassLoader`, which is what the third recommendation says in https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class

> Cache a reference to the ClassLoader object somewhere handy, and issue loadClass calls directly. This requires some effort.

My new solution is to actually go with the first recommendation with caching (caching is also better for performance as mentioned in https://developer.android.com/training/articles/perf-jni#jclass,-jmethodid,-and-jfieldid).

> Do your FindClass lookups once, in JNI_OnLoad, and cache the class references for later use. Any FindClass calls made as part of executing JNI_OnLoad will use the class loader associated with the function that called System.loadLibrary (this is a special rule, provided to make library initialization more convenient). If your app code is loading the library, FindClass will use the correct class loader.